text/0000-assoc-default-groups.md

```diff
@@ -0,0 +1,1104 @@
+and so it must be `42u8`, which has the type `u8`.
+Therefore, we have reached an inconsistency in the type system:
+`<Bar as Foo>::QUUX` is of value `42u8`, but of type `Vec<u8>`.
+So we may accept either `impl Foo for Bar` as defined in (4),
+or the definition of `Foo` as in (2), but not *both*.
+
+[RFC 192] solved this dilemma by rejecting the implementation
+and insisting that if you override *one* associated type,
+then you must override *all* other defaulted items.
+Or stated in its own words:
+
+> + If a trait implementor overrides any default associated types,
+>   they must also override all default functions and methods.
+> + Otherwise, a trait implementor can selectively override individual
+>   default methods/functions, as they can today.
+
+Meanwhile, as we saw in the error message above (3),
+the current implementation takes the alternative approach of accepting
+`impl Foo for Bar` (4) but not the definition of `Foo` as in (2).
+
+### Changes in this RFC
+[changes]: #changes-in-this-rfc
+
+In this RFC, we change the approach in [RFC 192] to the currently implemented
+approach. Thus, you will continue to receive the error message above
+and you will be able to provide associated type defaults.
+
+[specialization]: https://github.com/rust-lang/rfcs/pull/1210
+
+With respect to [specialization], the behaviour is the same.
+That is, if you write (5):
+
+```rust
+#![feature(specialization)]
+
+trait Foo {
+    type Bar;
+
+    fn quux(x: Self::Bar) -> u8;
+}
+
+struct Wibble<T>;
+
+impl<T> Foo for Wibble<T> {
+    default type Bar = u8;
+
+    default fn quux(x: Self::Bar) -> u8 { x }
+}
+```
+
+The compiler will reject this because you are not allowed to assume,
+just like before, that `x: u8`. The reason why is much the same as
+we have previously discussed in the [background].
+
+[current_impl_diverge]: https://play.rust-lang.org/?gist=30e01d77f7045359e30c7d3f3144e984&version=nightly&mode=debug&edition=2015
+
+One place where this proposal diverges from what is currently implemented
+is with respect to the [following example][current_impl_diverge] (6):
+
+```rust
+#![feature(associated_type_defaults)]
+
+trait Foo {
+    type Bar = usize;
+
+    fn baz(x: Self::Bar) -> usize;
+}
+
+impl<T> Foo for Vec<T> {
+    fn baz(x: Self::Bar) -> usize { x }
+}
+```
+
+In the current implementation, (6) is rejected because the compiler will not
+let you assume that `x` is of type `usize`. But in this proposal, you would be
+allowed to assume this. To permit this is not a problem because `Foo for Vec<T>`
+is not further specializable since `baz` in the implementation has not been
+marked as `default`.
+
+With these changes,
+we consider the design of associated type defaults to be *finalized*.
+
+## `default` specialization groups
+[default_groups]: #default-specialization-groups
+
+Note: Everything in this section assumes actual support for [specialization].
+
+Now, you might be thinking: - *"Well, what if I __do__ need to assume that
```

Comment:
This question made me think that `default` specialization groups is only useful if said group contains an associated type. This is then never challenged by having all examples have associated types in them.